File: separate-components.py
#!/usr/bin/env python3
"""
Separate Angular component templates and styles from TypeScript files.
This script reads component files and extracts template and styles to separate files.
"""

import os
import re
import sys
from pathlib import Path

# Get all .component.ts files
frontend_path = Path("/Users/mujthabamk/Desktop/real-world-projects/osa/OSA/osa-frontend")
components = list(frontend_path.rglob("*.component.ts"))

print(f"Found {len(components)} component files")

for component_file in components:
    # Skip if separate files already exist
    html_file = component_file.with_suffix('.html').with_stem(component_file.stem.replace('.component', '.component'))
    html_file = component_file.parent / f"{component_file.stem.split('.')[0]}.component.html"
    css_file = component_file.parent / f"{component_file.stem.split('.')[0]}.component.css"
    
    # Skip app.component and course-details which we already did
    if "app.component" in str(component_file) or "course-details" in str(component_file):
        print(f"✓ Skipping {component_file.name} (already processed)")
        continue
    
    # Read the original file
    content = component_file.read_text(encoding='utf-8')
    
    # Check if it already uses external files
    if 'templateUrl' in content or 'styleUrl' in content:
        print(f"✓ Skipping {component_file.name} (already using external files)")
        continue
    
    # Extract template
    template_match = re.search(r"template:\s*`([^`]*)`", content, re.DOTALL)
    if not template_match:
        # Try single-quoted template
        template_match = re.search(r"template:\s*'([^']*)'", content, re.DOTALL)
    
    # Extract styles
    styles_match = re.search(r"styles:\s*\[\s*`([^`]*)`\s*\]", content, re.DOTALL)
    if not styles_match:
        # Try with multiple styles
        styles_match = re.search(r"styles:\s*\[(.*?)\]", content, re.DOTALL)
    
    if template_match or styles_match:
        print(f"\nProcessing: {component_file.name}")
        
        # Write HTML file if template exists
        if template_match:
            template_content = template_match.group(1)
            html_file.write_text(template_content, encoding='utf-8')
            print(f"  ✓ Created {html_file.name}")
        
        # Write CSS file if styles exist
        if styles_match:
            styles_content = styles_match.group(1)
            # Clean up the styles content
            styles_content = re.sub(r"^\s*`\s*", "", styles_content)
            styles_content = re.sub(r"\s*`\s*,?\s*$", "", styles_content)
            css_file.write_text(styles_content, encoding='utf-8')
            print(f"  ✓ Created {css_file.name}")
        
        # Update the TS file
        new_content = content
        if template_match:
            template_start = content.find(template_match.group(0))
            template_end = template_start + len(template_match.group(0))
            new_content = new_content[:template_start] + f"templateUrl: '{html_file.name}'," + new_content[template_end:]
        
        if styles_match:
            styles_start = new_content.find(styles_match.group(0))
            styles_end = styles_start + len(styles_match.group(0))
            new_content = new_content[:styles_start] + f"styleUrl: '{css_file.name}'," + new_content[styles_end:]
        
        # component_file.write_text(new_content, encoding='utf-8')
        # print(f"  ✓ Updated {component_file.name}")
    else:
        print(f"⊘ No template or styles found in {component_file.name}")

print("\n✓ Script completed!")

File: separate-components.js
const fs = require('fs');
const path = require('path');

const FRONTEND_DIR = '/Users/mujthabamk/Desktop/real-world-projects/osa/OSA/osa-frontend';

function findComponentFiles(dir) {
  let files = [];
  const items = fs.readdirSync(dir);
  
  for (const item of items) {
    const fullPath = path.join(dir, item);
    const stat = fs.statSync(fullPath);
    
    if (stat.isDirectory()) {
      files = files.concat(findComponentFiles(fullPath));
    } else if (item.endsWith('.component.ts')) {
      files.push(fullPath);
    }
  }
  
  return files;
}

function extractTemplate(content) {
  // Match template with backticks
  const match1 = content.match(/template:\s*`([^`]*)`/s);
  if (match1) return match1[1];
  
  // Match template with single quotes
  const match2 = content.match(/template:\s*'([^']*)'(?=\s*[,}])/s);
  if (match2) return match2[1];
  
  return null;
}

function extractStyles(content) {
  // Match styles with backticks
  const match1 = content.match(/styles:\s*\[\s*`([^`]*)`\s*\]/s);
  if (match1) return match1[1];
  
  // More complex styles extraction
  const match2 = content.match(/styles:\s*\[([\s\S]*?)\]\s*[,}]/m);
  if (match2) {
    let styles = match2[1];
    // Remove backticks and leading/trailing whitespace
    styles = styles.replace(/^\s*`/, '').replace(/`\s*,?$/, '');
    if (styles.trim()) return styles;
  }
  
  return null;
}

function removeTemplateAndStyles(content, template, styles) {
  let result = content;
  
  if (template) {
    // Remove template from decorator
    result = result.replace(/template:\s*`[^`]*`/s, '');
    result = result.replace(/template:\s*'[^']*'/s, '');
  }
  
  if (styles) {
    // Remove styles from decorator
    result = result.replace(/styles:\s*\[[\s\S]*?\]/s, '');
  }
  
  // Clean up any double commas or trailing commas
  result = result.replace(/,\s*,/g, ',');
  result = result.replace(/,(\s*[}\]])/g, '$1');
  
  return result;
}

const components = findComponentFiles(path.join(FRONTEND_DIR, 'src'));
console.log(`\n📊 Found ${components.length} component files\n`);

let processed = 0;
let skipped = 0;

for (const componentFile of components) {
  const content = fs.readFileSync(componentFile, 'utf-8');
  const dir = path.dirname(componentFile);
  const baseName = path.basename(componentFile).replace('.component.ts', '');
  const htmlFile = path.join(dir, `${baseName}.component.html`);
  const cssFile = path.join(dir, `${baseName}.component.css`);
  
  // Skip if already using templateUrl or styleUrl
  if (content.includes('templateUrl') || content.includes('styleUrl')) {
    console.log(`⊘ ${baseName}.component.ts (already using external files)`);
    skipped++;
    continue;
  }
  
  const template = extractTemplate(content);
  const styles = extractStyles(content);
  
  if (!template && !styles) {
    console.log(`⊘ ${baseName}.component.ts (no template or styles)`);
    skipped++;
    continue;
  }
  
  console.log(`📦 Processing ${baseName}.component.ts`);
  
  if (template) {
    fs.writeFileSync(htmlFile, template, 'utf-8');
    console.log(`   ✓ Created ${baseName}.component.html (${template.length} bytes)`);
  }
  
  if (styles) {
    fs.writeFileSync(cssFile, styles, 'utf-8');
    console.log(`   ✓ Created ${baseName}.component.css (${styles.length} bytes)`);
  }
  
  processed++;
}

console.log(`\n✅ Completed!`);
console.log(`   Processed: ${processed} components`);
console.log(`   Skipped: ${skipped} components\n`);
